Mathlib/Data/List/Defs.lean

```diff
@@ -749,4 +749,72 @@ Example: if `f : ℕ → list ℕ → β`, `List.mapWithComplement f [1, 2, 3]`
 def mapWithComplement {α β} (f : α → List α → β) : List α → List β :=
   mapWithPrefixSuffix fun pref a suff => f a (pref ++ suff)
 
+/--
+Auxiliary function for `List.compareWith`
+-/
+def compareWithAux [Ord α] : List α → List α → Ordering
+  | x::xs, y::ys =>
+    let c := compare x y
+    if c == Ordering.eq then
+      compareWithAux xs ys
+    else c
+  | [], [] => Ordering.eq
+  | _, [] => Ordering.gt
+  | [], _ => Ordering.lt
+
+/--
+`List.compareWith` compares two lists `l₁` and `l₂` using thee function `compare` from
+the `Ord` typeclass, which means that the type of the list's elements have to have
+an instance of this typeclass. `List.comparewith` returns a contructor from the
+`Ordering` inductive type.
+
+Examples:
+```
+[1, 5].compareWith compare' [1, 8] = Ordering.lt
+[1, 10].compareWith compare' [1, 10] = Ordering.eq
+[1, 11].compareWith compare' [1, 13] = Ordering.gt
+[1, 5].compareWith compare' [1] = Ordering.lt
+[1].compareWith compare' [1, 8] = Ordering.gt
+```
+-/
+def compareWith [Ord α] (l₁ l₂ : List α) : Ordering :=
+  compareWithAux l₁ l₂
+
+/--
+`List.sum` sums all of the elements of of list `l`, whose elements
+have instances of `Add` and `OfNat`.
+-/
+def sum [Add α] [OfNat α 0] (l : List α) : α :=
+  l.foldl (.+.) 0
+
+/--
+`List.sumBy` sums the results of `f` mapped to each element of `l`. The return type of `f`
+must have instances of `Add` and `OfNat`.
+-/
+def sumBy [Add β] [OfNat β 0] (l : List α) (f : α → β) : β :=
+  l.map f |>.sum
+
+/--
+Allows `List.average` and `List.averageBy` to take `Float` as `α`s type,
+not restricting it to `Nat`.
+-/
+instance : HDiv Float Nat Float where
```

Comment:
Please remove the `HDiv` instance, we definitely don't want all combinations of `HDiv * * *` for all numeric types.